core/server/src/main/java/alluxio/worker/block/evictor/LIRSEvictor.java

```diff
@@ -0,0 +1,527 @@
+   * until LIR size is cut under the limit. The bottom of LIR cache on each StorageDir will never be
+   * a ghost block or HIR block. The bottom of LIR cache on each StorageDir is guaranteed to be a
+   * LIR block to mark the maximum recency of all IRR blocks so that all blocks in LIR cache has
+   * smaller recency than the maximum recency of IRR blocks and are possible to be transformed to a
+   * LIR block when they are accessed again.
+   *
+   * @param location location of the StorageDir
+   * @throws BlockDoesNotExistException if the meta data of the block can not be found
+   */
+  private void updateSizeOfLIR(BlockStoreLocation location) {
+    SpaceContainer spaceContainer = mSpaceManager.get(location);
+    Iterator<Map.Entry<Pair<BlockStoreLocation, Long>, BlockType>> it =
+        mLIRCache.entrySet().iterator();
+    if (!it.hasNext()) {
+      return;
+    }
+    Entry<Pair<BlockStoreLocation, Long>, BlockType> entry = it.next();
+    BlockType blockType = entry.getValue();
+    while (spaceContainer.getLIRBytes() > spaceContainer.getCapacity() * mLIRPercent
+        || !entry.getKey().getFirst().equals(location) || blockType.isHIR_Resident()
+        || !blockType.isResident()) {
+      if (!entry.getKey().getFirst().equals(location)) {
+        if (it.hasNext()) {
+          entry = it.next();
+          blockType = entry.getValue();
+          continue;
+        } else {
+          break;
+        }
+      } else if (blockType.isHIR_Resident() || !blockType.isResident()) {
+        it.remove();
+      } else if (blockType.isLIR()) {
+        long blockSize = mBlockIdToSize.get(entry.getKey().getSecond());
+        spaceContainer.moveBlockFromLIRToHIR(blockSize);
+        mHIRCache.put(entry.getKey(), BlockType.HIR_RESIDENT);
+        it.remove();
+      }
+      if (it.hasNext()) {
+        entry = it.next();
+        blockType = entry.getValue();
+      } else {
+        break;
+      }
+    }
+    mSpaceManager.put(location, spaceContainer);
+  }
+
+  /**
+   * Block types stored in LIR LRU stack.
+   */
+  enum BlockType {
+    // Blocks resident in HIR cache.
+    HIR_RESIDENT(1),
+    // Blocks moved from other tiers. Once accessed, they will be transformed to HIR_RESIDENT.
+    HIR_MOVED(2),
+    // Blocks resident in LIR cache.
+    LIR_RESIDENT(3),
+    // Blocks not resident on current tier.
+    EVICTED(4);
+    private final int mValue;
+
+    BlockType(int value) {
+      mValue = value;
+    }
+
+    public boolean isMoved() {
+      return mValue == HIR_MOVED.mValue;
+    }
+
+    public boolean isHIR_Resident() {
+      return mValue == HIR_RESIDENT.mValue;
+    }
+
+    public boolean isLIR() {
+      return mValue == LIR_RESIDENT.mValue;
+    }
+
+    public boolean isResident() {
+      return !(mValue == EVICTED.mValue);
+    }
+  }
+
+  /**
+   * Class to record the space used condition of a StorageDir.
+   */
+  class SpaceContainer {

```

Comment:
@uronce-cc  mHIRBytes + mLIRBytes != mCapacity. `SpaceContainer` holds used bytes of HIR cache and LIR cache.